src/rt/backtrace/dwarf.d

```diff
@@ -0,0 +1,464 @@
+{
+    import core.stdc.stdio : snprintf;
+
+    static if (includeBacktraceSymbols)
+    {
+        import core.sys.linux.execinfo : backtrace_symbols;
+        import core.sys.posix.stdlib : free;
+
+        const framelist = backtrace_symbols(callstack.ptr, cast(int) callstack.length);
+        scope(exit) free(cast(void*) framelist);
+    }
+
+    bool canResolveAddrToLine = false;
+    ElfSection dbgSection;
+
+    ElfFile file;
+    if (ElfFile.openSelf(&file))
+    {
+        auto stringSectionHeader = ElfSectionHeader(&file, file.ehdr.e_shstrndx);
+        auto stringSection = ElfSection(&file, &stringSectionHeader);
+
+        auto dbgSectionIndex = findSectionByName(&file, &stringSection, ".debug_line");
+        if (dbgSectionIndex != -1)
+        {
+            auto dbgSectionHeader = ElfSectionHeader(&file, dbgSectionIndex);
+            dbgSection = ElfSection(&file, &dbgSectionHeader);
+            canResolveAddrToLine = true;
+        }
+    }
+
+    int ret = 0;
+    foreach(size_t i; framesToSkip .. callstack.length)
+    {
+        // currently addresses are resolved one by one
+        // there is an optimization that allows resolving
+        // all at once if they are sorted
+
+        char[1536] buffer; buffer[0] = 0;
+        char[256] addressBuffer; addressBuffer[0] = 0;
+
+        if (canResolveAddrToLine)
+        {
+            size_t[1] addresses;
+            addresses[0] = cast(size_t) callstack[i];
+
+            resolveAddresses(&dbgSection, addresses[],
+                (Location location)
+                {
+                    snprintf(addressBuffer.ptr, addressBuffer.length, "@ %.*s:%d", cast(int) location.file.length, location.file.ptr, location.line);
+                }
+            );
+
+            if (addressBuffer[0] == '\0')
+                addressBuffer[] = "@ ??:?\0";
+        }
+
+        static if (includeBacktraceSymbols)
+        {
+            char[1024] symbolBuffer;
+            auto symbol = getDemangledSymbol(framelist[i][0 .. strlen(framelist[i])], symbolBuffer);
+            if (symbol.length > 0)
+                snprintf(buffer.ptr, buffer.length, "[0x%x] %.*s %s", callstack[i], cast(int) symbol.length, symbol.ptr, addressBuffer.ptr);
+            else
+                snprintf(buffer.ptr, buffer.length, "[0x%x] %s", callstack[i], addressBuffer.ptr);
+        }
+        else
+        {
+            snprintf(buffer.ptr, buffer.length, "[0x%x] %s", callstack[i], addressBuffer.ptr);
+        }
+
+        auto pos  = i - framesToSkip;
+        auto output = buffer[0 .. strlen(buffer.ptr)];
+        ret = dg(pos, output);
+        if (ret) break;
+    }
+
+    return ret;
+}
+
+private:
+
+// addresses must be sorted
+void resolveAddresses(ElfSection* debugLineSection, const(size_t)[] addresses, scope void delegate(Location location) callback)
```

Comment:
It can be simplified into `resolveAddress` for the time being as this is how it is used currently. I think I'll do this.